File: go/fibonacci/fibonacci.go
// Fibonacci Series Calculator
//
// The Fibonacci Series Calculator is a Go program that calculates and
// prints the Fibonacci series up to the nth term with arbitrary
// precision. It includes features such as error handling for input
// validation, handling special cases where numbers don't end in "th"
// (e.g. 1, 2, 3), displaying the sum of the series, and providing the
// current date and time.
//
// License: This program is in the public domain.

package main

import (
	"bufio"
	"fmt"
	"io"
	"math/big"
	"os"
	"os/signal"
	"strings"
	"time"
)

func main() {
	fmt.Println("============Fibonacci Series Calculator============")
	fmt.Println("This Program was Written Using: Go")

	signalHandler()
	dateAndTime()
	getUserInput()

	fmt.Println("===================================================")
}

// Function to Get the User Input
func getUserInput() {
	reader := bufio.NewReader(os.Stdin)

	for {
		fmt.Print("Enter the value of n (an integer): ")
		input, err := reader.ReadString('\n')

		if err == io.EOF {
			eofHandler()
		}

    trimmedInput := strings.TrimSpace(input)

    if trimmedInput == "" {
			fmt.Println("Please enter something...")
			continue
		} else if trimmedInput == "exit" {
			fmt.Println("Exiting the program...")
			break
		}

		n := validateInput(trimmedInput)

		if n > 0 {
			fibonacciSeries(n)
			break
		} else {
			fmt.Println("Please enter a valid positive integer.")
		}
	}
}

// Calculates and prints the Fibonacci series up to the nth term.
func fibonacciSeries(n int) {
	a := new(big.Int)
	b := new(big.Int)
	temp := new(big.Int)
	sum := new(big.Int)

	a.SetUint64(0)
	b.SetUint64(1)

	fmt.Printf("Fibonacci series up to the %d%s term:\n", n, getSuffix(n))

	var series []string

	for i := 0; i <= n; i++ {
    if n <= 5000 {
		  series = append(series, a.String())
    } else {
      // Print the series without using array
      fmt.Print(a.String())
      if i < n {
        fmt.Print(", ")
      } else {
        fmt.Println("")
      }
    }

		temp.Set(a)
		a.Set(b)
		b.Add(b, temp)
		sum.Add(sum, temp) // Calculate the sum
	}
  if n <= 5000 {
	  fmt.Println(strings.Join(series, ", "))
  }

  fmt.Println("")
	fmt.Println("Sum of the Fibonacci series:", sum.String())
}

// Handle special cases where numbers don't end in "th"
func getSuffix(n int) string {
	if n%10 == 1 && n%100 != 11 {
		return "st"
	} else if n%10 == 2 && n%100 != 12 {
		return "nd"
	} else if n%10 == 3 && n%100 != 13 {
		return "rd"
	} else {
		return "th"
	}
}

func dateAndTime() {
	currentDate := time.Now()

	format := "January 02, 2006 - 15:04:05"
	formattedDate := currentDate.Format(format)

	fmt.Println("Date and Time:", formattedDate)
}

// Functions to handles user input and errors
func validateInput(input string) int {
	n := new(big.Int)
	n, ok := n.SetString(input, 10)
	if !ok || n.Cmp(big.NewInt(0)) < 1 {
		return -1
	}
	return int(n.Int64())
}

func eofHandler() {
	fmt.Println()
	fmt.Println("End of File encountered.. Stopping...")
	fmt.Println("===================================================")
	os.Exit(0)
}

func interruptHandler(callback func()) {
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt)
	go func() {
		<-c
		fmt.Println()
		fmt.Println("Interrupt received.. Exiting...")
		fmt.Println("===================================================")
		callback()
		os.Exit(0)
	}()
}

func signalHandler() {
  interruptHandler(func() {})
}

File: rust/fibonacci/src/fibonacci.rs
/// Fibonacci Series Calculator
///
/// The Fibonacci Series Calculator is a Rust program that calculates
/// and prints the Fibonacci Series up to the nth term with arbitrary
/// precision. It includes features such as error handling for input
/// validation, handling special cases where numbers don't end in "th"
/// (e.g. 1, 2, 3), displaying the sum of the series, and providing
/// the current date and time.
///
/// License: This program is in the public domain.

extern crate chrono;
extern crate ctrlc;

use chrono::Local;
use lazy_static::lazy_static;
use num_bigint::BigUint;
use std::io::{self, Write};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

fn main() {
  println!("============Fibonacci Series Calculator============");
  println!("This Program was Written Using: Rust");

  signal_handler();
  date_and_time();
  get_user_input();

  println!("===================================================");
}

/// Function to get user input.
fn get_user_input() {
  loop {
    print!("Enter the value of n (an integer): ");
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();

    if input.is_empty() {
      eof_handler();
    }

    let trimmed_input = input.trim();

    if trimmed_input.is_empty() {
      println!("Please enter something...");
      continue;
    } else if trimmed_input == "exit" {
      println!("Exiting the program...");
      break;
    }

    match validate_input(&trimmed_input) {
      Some(n) => {
        fibonacci_series(n);
        break;
      }
      None => {
        println!("Please enter a valid positive integer.");
      }
    }
  }
}

/// Calculates and prints the Fibonacci Series up to the nth term.
fn fibonacci_series(n: u32) {
  let mut a: BigUint = 0u32.into();
  let mut b: BigUint = 1u32.into();
  let mut temp: BigUint;
  let mut sum: BigUint = 0u32.into();

  println!("Fibonacci Series up to the {}{} term:", n, get_suffix(n));

  let mut series: Vec<BigUint> = Vec::new();

  for i in 0..=n {
    if n <= 5000 {
      series.push(a.clone());
    } else {
      // Print the series without using arrays
      print!("{}", a);
      if i < n {
        print!(", ");
      }
    }

    temp = a.clone();
    a = b.clone();
    b += temp.clone();
    sum += temp; // Calculate the sum
  }

  if n <= 5000 {
    let fib_series: String = series
      .iter()
      .map(|x| x.to_string())
      .collect::<Vec<String>>()
      .join(", ");
    print!("{}", fib_series);
  }

  println!("\n");
  println!("Sum of the Fibonacci Series: {}", sum);
}

/// Handle special cases where numbers don't end in "th"
fn get_suffix(n: u32) -> &'static str {
  if n % 10 == 1 && n % 100 != 11 {
    "th"
  } else if n % 10 == 2 && n % 100 != 12 {
    "nd"
  } else if n % 10 == 3 && n % 100 != 13 {
    "rd"
  } else {
    "th"
  }
}

/// Display the current date and time
fn date_and_time() {
  let current_date = Local::now();

  let layout = "%B %d, %Y - %H:%M:%S";
  let formatted_date = current_date.format(layout).to_string();

  println!("Date and Time: {}", formatted_date);
}

/// Functions to handle user input and errors
fn validate_input(input: &str) -> Option<u32> {
  match input.parse() {
    Ok(n) if n > 0 => Some(n),
    _ => None,
  }
}

fn eof_handler() {
  println!();
  println!("End of File encountered.. Stopping...");
  println!("===================================================");
  std::process::exit(0);
}

lazy_static! {
  static ref RUNNING: Arc<AtomicBool> = Arc::new(AtomicBool::new(true));
  static ref RUN: Arc<AtomicBool> = Arc::clone(&RUNNING);
}

fn interrupt_handler() {
  println!();
  println!("Interrupt received.. Exiting...");
  println!("===================================================");
  std::process::exit(0);
}

fn signal_handler() {
  ctrlc::set_handler(move || {
    interrupt_handler();
    RUN.store(false, Ordering::SeqCst);
  })
  .expect("Error setting SIGINT handler.");
}
